[Parser][QoI] Offer fixit for changing the platform condition kind

roop opened the pull request:
Resolves [SR-11037](https://bugs.swift.org/browse/SR-11037).

In case there’s an exact match with a value of an alternate kind, this solution offers a single fixit to change the platform conditional kind. For example:

```
#if os(simulator)
// Fixit: os -> targetEnvironment
#endif
```

In case there’s only a case-insensitive match with a value of an alternate kind, this solution offers two fixits: one for the kind and one for the value. For example:

```
#if os(Simulator)
// Fixit: os -> targetEnvironment
// Fixit: Simulator -> simulator
#endif
```
Implementation details:

To search for a match in an alternate platform conditional kind, we need a way to loop through the different platform conditional value arrays (SupportedConditionalCompilationOSs, SupportedConditionalCompilationArches, …), for which I’m using a std::pair<const StringRef*, size_t> to represent each array. I’m open to suggestions on an alternate representation.

roop committed:
[Parser][QoI] Offer fixit for changing the platform condition kind

For example, for "#if os(simulator)", offer a fixit to change
"os" to "targetEnvironment", instead of offering to change "simulator".

Resolves SR-11037.

theblixguy commented:
cc @brentdax

brentdax commented:
@swift-ci please test

(Sorry about the delay—I've been a little busy.)

brentdax commented on lib/Parse/ParseIfConfig.cpp:
```diff
@@ -47,6 +47,24 @@ Optional<PlatformConditionKind> getPlatformConditionKind(StringRef Name) {
 }
 
+/// Get platform condition name from PlatformConditionKind.
+static StringRef getPlatformConditionName(PlatformConditionKind Kind) {
```
If we're going this far, we should probably create a `PlatformConditionKind.def` file and include it here and in the other places where we list out all the platform conditions. [Here's a commit from a few months ago where I added a similar .def file.](https://github.com/apple/swift/commit/31aeb64efa23cedb7ea66b45f7ff7bcfa5a1a4c1) In this case, I think you'll want the entries to look something like:

```
#ifndef PLATFORM_CONDITION
#define PLATFORM_CONDITION(LABEL, IDENTIFIER)
#endif
#ifndef PLATFORM_CONDITION_
#define PLATFORM_CONDITION_(LABEL, IDENTIFIER) PLATFORM_CONDITION(LABEL, "_" ## IDENTIFIER)
#endif

PLATFORM_CONDITION(OS, "os")
PLATFORM_CONDITION(Arch, "arch")
PLATFORM_CONDITION_(Endianness, "endian")
PLATFORM_CONDITION_(Runtime, "runtime")
PLATFORM_CONDITION(CanImport, "canImport")
PLATFORM_CONDITION(TargetEnvironment, "targetEnvironment")

#undef PLATFORM_CONDITION
#undef PLATFORM_CONDITION_
```

(The separate `PLATFORM_CONDITION_` definition lets you treat the internal and public conditions differently.)

roop replied:
Ok, got it.

brentdax commented on lib/Basic/LangOptions.cpp:
```diff
@@ -85,28 +143,21 @@ bool contains(const StringRef (&Array)[N], const StringRef &V,
 checkPlatformConditionSupported(PlatformConditionKind Kind, StringRef Value,
-                                std::vector<StringRef> &suggestions) {
+                                PlatformConditionKind &suggestedKind,
+                                std::vector<StringRef> &suggestedValues) {
```
Would it make sense to return the suggestions as a `std::vector<std::pair<PlatformConditionKind, StringRef>> &` instead? That way we could suggest more flexible corrections as conditions warrant.

roop replied:
I don't think **[(PlatformConditionKind, StringRef)]** would fit well here. Sometimes, we want multiple value suggestions for the same PlatformConditionKind (for e.g. `#if os(chaos)` should suggest multiple closest-edit-distance-matching values for the "os" kind), and sometimes, we want no value suggestions for a PlatformConditionKind (for e.g. `#if os(simulator)` should suggest only a kind change, not a value change).

Are you saying that maybe in the future, we might want to return multiple PlatformConditionKind suggestions? In that case, maybe **[(PlatformConditionKind, [StringRef])]** would be more appropriate, because we can have 0..<n value suggestions for a given kind. Nevertheless, I don't think that change should be made now -- it should rather be done when we do start returning multiple PlatformConditionKind suggestions. What do you think?

brentdax replied:
That's fair.

brentdax commented on test/Parse/ConditionalCompilation/basicParseErrors.swift:
```diff
@@ -139,6 +139,24 @@ func undefinedFunc() // ignored.
+#if _endian(arm64) // expected-warning {{unknown endianness for build configuration '_endian'}} expected-note {{did you mean 'arch'}} {{5-12=arch}}
+#endif
+
+#if targetEnvironment(_ObjC) // expected-warning {{unknown target environment for build configuration 'targetEnvironment'}} expected-note {{did you mean '_runtime'}} {{5-22=_runtime}}
```
Come to think of it, we probably *shouldn't* suggest underscored build configurations—they're not officially supported, after all.

roop replied:
Makes sense. This is addressed in commit [e44a458](https://github.com/apple/swift/pull/25977/commits/e44a458ec8df83275426b1db4c828439b53325c1).

beccadax reviewed (changes requested):
I like where this is going!

roop committed:
Make a .def file for PlatformConditionKind

roop commented:
I’m afraid I will not be working on this for a few more days. I hope to continue working on this on Tue 16/Jul. Sorry about the delay.

brentdax commented:
@roop I started as a volunteer contributor—I know how it is. Thanks for tackling this!

roop committed:
[Parser][QoI] Suppress fixits to underscored platform condition kinds

"_endian" and "_runtime" aren't officially supported platform condition
kinds, so don't suggest changing to one of those.

brentdax commented on include/swift/AST/PlatformConditionKinds.def:
```diff
@@ -0,0 +1,33 @@
+//===--- ObjCSelectorFamily.def - Objective-C Selector Families - C++ ---*-===//
```
Update the filename here.

roop replied:
Sorry about that. :(

brentdax replied:
I've done much worse! 🤣

brentdax commented on include/swift/Basic/LangOptions.h:
```diff
@@ -37,18 +37,8 @@ namespace swift {
 
   /// Kind of implicit platform conditions.
   enum class PlatformConditionKind {
-    /// The active os target (OSX, iOS, Linux, etc.)
```
Can you move these comments into PlatformConditionKinds.def?

brentdax commented on lib/Basic/LangOptions.cpp:
```diff
@@ -65,15 +65,70 @@ static const StringRef SupportedConditionalCompilationTargetEnvironments[] = {
   "simulator",
 };
 
+static const PlatformConditionKind AllPublicPlatformConditionKinds[] = {
```
Excellent.

beccadax reviewed (changes requested):
Much better! Just a couple of nitpicks.

roop commented:
Ideally, the previous 2 commits should've been an update to the "Make a .def file for PlatformConditionKind" commit, but I'm not sure of the Swift project's policy on force-pushing to PRs under review, so I'm playing it safe and just adding commits on top.

brentdax commented:
@roop Force-pushes in PRs are just fine. They do confuse the CI infrastructure a bit, though—you get an "error" which is really just the job being rescheduled with a different commit hash.

brentdax commented:
@swift-ci please test

brentdax commented:
@swift-ci please test

roop commented:
@brentdax Can you please start the ci on this again?

I don't think the Linux failure was related to this change. I tried running smoke tests and validation tests (through lit.py) on Linux after merging to a recent master, and they both passed.

brentdax commented:
@swift-ci please test

roop commented:
The "Swift Test Linux Platform" and "Swift Test OS X Platform" both seem to have actually passed - I presume swift-ci reporting them as failures is a result of it getting confused by the force-push, as you said earlier.

The Linux smoke test says "Expected" -- maybe the test hasn't finished running yet?

The OS X smoke test seems to have passed as per swift-ci but I can't actually see the jenkins log (following the Details link says "Not Found").

So let's wait for the Linux smoke test. If that looks good, can we merge this?

brentdax commented:
There was an issue with llbuild and SwiftPM last night; it should now be resolved.

@swift-ci please test

beccadax reviewed (changes requested):
I'm sorry—I hate to do this, but there's one issue left that I didn't notice until now.

roop commented on lib/Basic/LangOptions.cpp:
```diff
@@ -65,15 +65,70 @@ static const StringRef SupportedConditionalCompilationTargetEnvironments[] = {
+  return N;
+}
+
+std::pair<const StringRef*, size_t> getSupportedConditionalCompilationValues(const PlatformConditionKind &Kind) {
```
Makes sense (I didn’t know about ArrayRef). No worries. Will make these changes tomorrow.

jrose-apple replied:
ArrayRef's elements are always immutable already, so no need for the `const`.

roop committed:
Simplify code using ArrayRef

roop commented:
Pushed commit cf9bb32 to address @brentdax's suggestion to use ArrayRef. (It's a force push because the old branch wasn't building with the latest llvm, so I rebased it onto master before making the changes.)

brentdax commented:
@swift-ci please test

brentdax commented:
Note in case the build czar needs it: Integration is tracked by rdar://54449041.

roop commented:
🎉 Thanks @brentdax.

brentdax commented:
@roop Thank you for doing the work! I just wish I'd gotten it in faster.
